core/corehttp/hostname.go

```diff
@@ -151,16 +159,44 @@ func HostnameOption() ServeOption {
 					return
 				}
 
-				// Do we need to fix multicodec in PeerID represented as CIDv1?
-				if isPeerIDNamespace(ns) {
-					keyCid, err := cid.Decode(rootID)
-					if err == nil && keyCid.Type() != cid.Libp2pKey {
-						if newURL, ok := toSubdomainURL(hostname, pathPrefix+r.URL.Path, r); ok {
-							// Redirect to CID fixed inside of toSubdomainURL()
+				// Check if rootID is a valid CID
+				if rootCID, err := cid.Decode(rootID); err == nil {
+					// Do we need to redirect root CID to a canonical DNS representation?
+					dnsCID, err := toDNSPrefix(rootID, rootCID)
+					if err != nil {
+						http.Error(w, err.Error(), http.StatusBadRequest)
+						return
+					}
+					if !strings.HasPrefix(r.Host, dnsCID) {
```

Comment:
What case is this meant to deal with?

Reply:
This is a failsafe to ensure that if (for some reason) requested CID is not the same as DNS-safe-version we redirect to DNS-safe-version.

This could occur if user opened `{ed25519-in-base32}.ipns.example.com` in a browser that uses Gateway port as HTTP Proxy. In that mode DNS resolution does not fail due to the 63char limit, because hostname is passed to HTTP Proxy (go-ipfs) as an opaque string (no DNS lookup), so entire thing will load just fine.

Why we care about normalization in this edge case? 

- Someone could copy `{ed25519-in-base32}` from address bar and share with a friend that does not use gateway in proxy mode, and that would produce DNS failure. By redirecting to `{ed25519-in-base36}` version we ensure consistency of Origins across IPNS websites loaded form the subdomain gw, including proxy modes.
- We use proxy mode in Firefox because of https://bugzilla.mozilla.org/show_bug.cgi?id=1220810 so this edge case could impact some users of IPFS Companion.